Add warning diagnostic for comma operator usage in indexing operation

yunusey opened the pull request:
This commit adds a new diagnostic to point out misleading use of comma operators in indexing operation (e.g., let b = a[i, j]).
Closes #939

yunusey commented:
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

yunusey commented:
Addresses the issue #939.

yunusey commented:
Closes #939

strager commented on docs/errors/E0450.md:
````diff
@@ -0,0 +1,20 @@
+let b = a[a[0], a[1], a[2]];
+```
+
+Introduced in quick-lint-js version 2.11.0.
````
Must fix: This is wrong. Version 2.11.0 has already been released.

I'd just delete this paragraph for now:
```suggestion
```

strager commented on src/quick-lint-js/fe/diagnostic-types.h:
```diff
@@ -558,6 +558,12 @@
+  QLJS_DIAG_TYPE(                                                               \
+      diag_misleading_comma_operator_in_index_operation, "E0450",               \
+      diagnostic_severity::warning, { source_code_span comma; },                \
+      MESSAGE(QLJS_TRANSLATABLE("misleading use of ',' operator in in index"),  \
```
Fix typo:
```suggestion
      MESSAGE(QLJS_TRANSLATABLE("misleading use of ',' operator in index"),  \
```

strager commented on docs/errors/E0450.md:
```diff
@@ -0,0 +1,20 @@
+# E0450: misleading use of comma operator in indexing operation
```
This doc title should match the diagnostic's message ("misleading use of ',' operator in in index"). Either change this title or change the diagnostic in diagnostic-types.h.

strager commented on test/test-parse-warning.cpp:
```diff
@@ -342,6 +342,86 @@ TEST_F(test_parse_warning,
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.errors, IsEmpty());
+  }
+}
```
Nitpick: I think this testing is excessive for how simple the implementation is. I'd deduplicate some tests, such as the `a[b[1967, 1975]]` and `a[0][-1, 1]` and `let b = a[-1, 1]` tests which cover the same thing.

strager commented on src/quick-lint-js/fe/diagnostic-types.h:
```diff
@@ -558,6 +558,12 @@
+  QLJS_DIAG_TYPE(                                                               \
+      diag_misleading_comma_operator_in_index_operation, "E0450",               \
+      diagnostic_severity::warning, { source_code_span comma; },                \
+      MESSAGE(QLJS_TRANSLATABLE("misleading use of ',' operator in in index"),  \
```
Minor: I think we should also point to the `[` in a secondary message. This could help people understand what 'index' means exactly.

    foo[bar, baz]
    //     ^ misleading use of ',' operator in index
    // ^ index starts here

yunusey replied:
I have added an additional diagnostic note as in the example above, and also addressed the other issues you pointed out. Thanks!

strager reviewed (changes requested):
Overall, this patch looks good. 👍

strager commented:
The 'check error docs' CI job detected some issues: https://github.com/quick-lint/quick-lint-js/actions/runs/4289586245/jobs/7477197134

The existing example in docs/errors/E0062.md is triggering for your code, but shouldn't. It looks like our error recovery is bad here:

https://github.com/quick-lint/quick-lint-js/blob/7863d029128d7562134171c7f54e152bf46b0d1c/src/quick-lint-js/fe/parse-statement.cpp#L1501-L1502

The problem is that `parse_expression_remainder` is called which parses `[` as an index operation. I think a good fix for this is to only call `parse_expression_remainder` if the current token is `(` (IIFE).

strager commented on docs/errors/E0450.md:
````diff
@@ -0,0 +1,18 @@
+If the intention is to create a new array `b` with the first three elements of array `a` the correct syntax would be:
+```javascript
+let a = [1, 2, 3, 4];
+let b = a[a[0], a[1], a[2]];
````
Must fix: This line triggers E0450. CI catches this mistake: https://github.com/quick-lint/quick-lint-js/actions/runs/4289586245/jobs/7477197134

> /__w/quick-lint-js/quick-lint-js/docs/errors/E0450.md: error: expected no error in code block #2 but found errors

I think you mean this?
```suggestion
let b = [a[0], a[1], a[2]];
```

Looking at it more closely, though, I think the comma syntax is more often used for multidimensional arrays, so I suggest you change your examples here:
```javascript
let matrix = [
  [1, 2, 3],
  [4, 5, 6],
  [7, 8, 9],
];
let topLeft = matrix[0, 0]; // should be: matrix[0][0]
```

yunusey replied:
Yes, I fixed that typo, and enhanced the error recovery around IIFE as you suggested.

yunusey commented:
I just saw that the CI running on the documentation failed again due to having both E0062 and E0061 in the documentation of E0062. But with the change we just made to better handle IIFE, I think this is expected, and we should remove the second half of the first block in [```docs/errors/E0062.md```.](https://github.com/quick-lint/quick-lint-js/blob/master/docs/errors/E0062.md) It is practically identical to the test we have in [```docs/errors/E0061.md```](https://github.com/quick-lint/quick-lint-js/blob/master/docs/errors/E0061.md), so should produce the same error.

strager commented:
> But with the change we just made to better handle IIFE, I think this is expected, and we should remove the second half of the first block in [docs/errors/E0062.md.](https://github.com/quick-lint/quick-lint-js/blob/master/docs/errors/E0062.md) It is practically identical to the test we have in [docs/errors/E0061.md](https://github.com/quick-lint/quick-lint-js/blob/master/docs/errors/E0061.md), so should produce the same error.

I agree!

Can you do this (and the related fix) in a separate PR? If you don't want to deal with the Git-fu, that's fine; you can put the fixes in this PR and I can split them out when I merge.

Unrelated: I fixed the 'test Emacs plugin' and 'test Nix install' issues in master, so if the failures bother you, you can rebase or merge master to get those fixed.

yunusey commented:
> 
> I agree!
> 
> Can you do this (and the related fix) in a separate PR? If you don't want to deal with the Git-fu, that's fine; you can put the fixes in this PR and I can split them out when I merge.
> 
> Unrelated: I fixed the 'test Emacs plugin' and 'test Nix install' issues in master, so if the failures bother you, you can rebase or merge master to get those fixed.

Hi @strager, I created a separate PR (#942) for IIFE related changes, and squashed all of my other changes into a single commit here after merging master. Hope I have not broken the git history :)

strager commented:
Landed as commit 35d94b9d6c65e6b645189ae22c918e1f6ac526db.
